### PR TITLE
[DNM] Fix type system and fix downgrade index scan

### PR DESCRIPTION
**Pull request by birdstorm**
DO NOT MERGE.
This PR will still be broken into separate PRs.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/execution/CoprocessorRDD.scala [hunk 1]
```diff
@@ -187,7 +187,6 @@ case class RegionTaskExec(child: SparkPlan,
   // cache invalidation call back function
   // used for driver to update PD cache
   private val callBackFunc = CacheInvalidateListener.getInstance()
-  private val downgradeSplitFactor = 4
```
Comment (marsishandsome) [inline core/src/main/scala/org/apache/spark/sql/execution/CoprocessorRDD.scala]: why remove this line?

File: core/src/main/scala/org/apache/spark/sql/execution/CoprocessorRDD.scala [hunk 2]
```diff
@@ -261,7 +260,7 @@ case class RegionTaskExec(child: SparkPlan,
         taskCount += 1
         val task = new Callable[util.Iterator[TiRow]] {
           override def call(): util.Iterator[TiRow] =
-            CoprocessIterator.getRowIterator(dagRequest.copy(), tasks, session)
+            CoprocessIterator.getRowIterator(dagRequest, tasks, session)
```
Comment (marsishandsome) [inline core/src/main/scala/org/apache/spark/sql/execution/CoprocessorRDD.scala]: why remove `copy`?

File: core/src/main/scala/com/pingcap/tispark/TiDBRelation.scala [hunk 3]
```diff
@@ -48,8 +48,11 @@ case class TiDBRelation(session: TiSession,
 
   override def sizeInBytes: Long = tableRef.sizeInBytes
 
-  def logicalPlanToRDD(dagRequest: TiDAGRequest): TiRDD =
-    new TiRDD(dagRequest, session.getConf, tableRef, session, sqlContext.sparkSession)
+  def logicalPlanToRDD(dagRequest: TiDAGRequest): TiRDD = {
+    val tiConf = session.getConf
+    tiConf.setTypeSystemVersion(sqlContext.sparkSession.conf.get(TiConfigConst.TYPE_SYSTEM_VERSION, "0").toInt)
```
Comment (marsishandsome) [inline core/src/main/scala/com/pingcap/tispark/TiDBRelation.scala]: why `setTypeSystemVersion` here?

File: core/src/main/scala/org/apache/spark/sql/execution/CoprocessorRDD.scala (hunk 2 above)
Comment (birdstorm) [reply core/src/main/scala/org/apache/spark/sql/execution/CoprocessorRDD.scala]: `copy()` is called in getRowIterator

File: core/src/main/scala/com/pingcap/tispark/TiDBRelation.scala (hunk 3 above)
Comment (zhexuany) [reply core/src/main/scala/com/pingcap/tispark/TiDBRelation.scala]: This allows changing conf dynamically

**Comment by birdstorm** [comment]
/run-all-tests tidb=release-2.1 tikv=release-2.1 pd=release-2.1

File: core/src/main/scala/com/pingcap/tispark/utils/TiUtil.scala [hunk 4]
```diff
@@ -111,17 +114,40 @@ object TiUtil {
                               blacklist: ExpressionBlacklist): Boolean =
     isSupportedBasicExpression(expr, source, blacklist) && isPushDownSupported(expr, source)
 
+  def toSparkRow(row: TiRow, rowTransformer: RowTransformer, version: Int): Row = {
```
Comment (ilovesoup) [inline core/src/main/scala/com/pingcap/tispark/utils/TiUtil.scala]: If you need pass version around, I'd rather make it an class with version as field.

File: core/src/main/spark-2.3/org/apache/spark/sql/catalyst/catalog/TiDirectExternalCatalog.scala [hunk 5]
```diff
@@ -38,7 +39,8 @@ class TiDirectExternalCatalog(tiContext: TiContext) extends ExternalCatalog {
 
   override def getTable(db: String, table: String): CatalogTable = {
     val schema = TiUtil.getSchemaFromTable(
-      meta.getTable(db, table).getOrElse(throw new NoSuchTableException(db, table))
+      meta.getTable(db, table).getOrElse(throw new NoSuchTableException(db, table)),
+      tiContext.conf.getInt(TiConfigConst.TYPE_SYSTEM_VERSION, 0)
```
Comment (ilovesoup) [inline core/src/main/spark-2.3/org/apache/spark/sql/catalyst/catalog/TiDirectExternalCatalog.scala]: If there are multiple places with default version 0, there should be an constant for default version. Also in TiDirectExternalCatalog.scala

File: tikv-client/src/main/java/com/pingcap/tikv/operation/SchemaInfer.java [hunk 6]
```diff
@@ -43,7 +43,7 @@ public static SchemaInfer create(TiDAGRequest dagRequest) {
   }
 
   public static SchemaInfer create(TiDAGRequest dagRequest, boolean readHandle) {
-    return new SchemaInfer(dagRequest, readHandle);
+    return new SchemaInfer(dagRequest.copy(), readHandle);
```
Comment (ilovesoup) [inline tikv-client/src/main/java/com/pingcap/tikv/operation/SchemaInfer.java]: @birdstorm make sure your change does not been override.

File: tikv-client/src/main/java/com/pingcap/tikv/util/RangeSplitter.java [hunk 7]
```diff
@@ -127,58 +123,26 @@ private RangeSplitter(RegionManager regionManager) {
     TLongObjectHashMap<TLongArrayList> result = new TLongObjectHashMap<>();
     handles.sort();
 
-    int startPos = 0;
-    DecodeResult decodeResult = new DecodeResult();
-    while (startPos < handles.size()) {
-      long curHandle = handles.get(startPos);
+    byte[] endKey = null;
```
Comment (ilovesoup) [inline tikv-client/src/main/java/com/pingcap/tikv/util/RangeSplitter.java]: Please make sure this one works fine. Check this with user who reported issue.

**Comment by birdstorm** [comment]
Should be dispersed into #730, #714, and type system fix. @zhexuany

**Pull request closed without merging**